[FEATURE] JWT 토큰 생성 로직 구현

JSY8869 opened the pull request:
## 제목
Jwt기반으로 실제 토큰을 생성하고 제공하는 로직 구현

## 작업 내용
1. 토큰 도메인 구현

토큰의 확장을 고려하여 Token을 AccessToken과 RefreshToken으로 분리하고 각각을 구현하여 사용하는 방식으로 구현

2. Jwt 토큰 생성 로직 구현

TokenProperties의 값과 실제 사용자의 정보를 이용하여 토큰을 생성하는 로직 구현

3. Jwt 토큰 제공 로직 구현

access토큰과 refresh토큰을 생성하여 JwtToken라는 하나의 객체로 제공하는 provider 구현

4. Jwt 토큰 서비스 로직 구현

서비스 로직을 provider와 creator에서 수행하기 때문에 특별한 내용 없이 추상화를 고려하여 구현

5. MainConfiguration에 configuration 추가

위의 객체들을 스프링 빈으로 등록하여 사용하기 위해 configuration 추가

6. JwtUserDetails 내의 MemoryUser의 정보를 얻기 위한 로직 추가

TokenCreator 내에서 JwtUserDetails의 MemoryUser를 꺼내서 값을 가져오는 방식이 아닌 JwtUserDetails에서 MemoryUser의 값을 TokenCreator에게 직접 넘겨주는 방식으로 변경

JSY8869 committed:
chore: add jjwt-api in build.gradle for use jjwt

1. build.gradle 파일 수정

jjwt 라이브러리를 사용하기 위해 build.gradle 파일에 추가하였습니다.

CokeLee777 committed:
refactor: refactoring UserDetails

UserDetails 클래스 도메인과 분리 -> 다형성 구현

related to #12

JSY8869 committed:
Merge branch 'feature/12' of https://github.com/CokeLee777/spring-security-jwt-auth into feature/12

JSY8869 committed:
feat: Add token domain

1. Access, Refresh 토큰을 각각 구현, 관리할 수 있도록 두 토큰이 따로 Token을 상속 받도록 설계하였습니다.

2. JwtAccessToken, JwtRefreshToken 에서 각각의 토큰을 구현한 후, 하나의 객체로 관리하기 위해 JwtToken 파일을 구현하였습니다.

3. domain 폴더 내에서 인터페이스와 구현체를 분리하기 위해 jwt 폴더를 생성한 후 jwt 관련 구현체를 해당 폴더에 저장하였습니다.

related to #12

JSY8869 committed:
feat: Add token creator

1. 각각의 용도에 맞게 token creator를 확장해서 구현할 수 있도록 인터페이스를 통해 추상화 하였습니다.

2. JwtTokenCreator에서 Jwt에 대한 TokenCreator를 구현하였으며 유저의 정보와 함께 어떤 토큰 인지에 대한 정보를 포함하도록 설계하였습니다.

3. 직접 입력하면서 생길 수 있는 사용자의 실수를 줄이고, 값들을 한 곳에서 관리하기 위해 TokenProperties 파일을 생성하여 해당 파일에서 Token의 설정 값들을 관리하도록 설계하였습니다.

related to #12

JSY8869 committed:
feat: Add token provider

1. token creator에게 단일 책임 원칙을 지키기 위해 creator와 provider를 분리하여 각각 생성과 제공을 담당하도록 설계하였습니다.

2. access 토큰과 refresh 토큰을 하나의 객체로 반환하기 위해 JwtToken의 객체를 생성하여 반환하도록 설계하였습니다.

related to #12

JSY8869 committed:
feat: Add token service

1. token 서비스를 인터페이스로 추상화 하여 다른 토큰의 서비스를 확장시켜 구현할 수 있도록 설계하였습니다.

related to #12

JSY8869 committed:
feat: Add jwt token configuration

1. Jwt 토큰 관련 객체들을 스프링 빈으로 등록하여 사용하기 위해 configuration 파일에 등록하였습니다.

related to #12

JSY8869 committed:
fix: Add JwtMemoryUserDetails getter

1. JwtTokenCreator 에서 JwtMemoryUserDetails의 객체를 통해 유저의 Id, nickname 등을 가져오기 위해 memoryuser 내부의 값을 반환하는 메서드를 추가하였습니다.

2. 책임과 역할을 분리하기 위해 JwtTokenCreator에서 직접 MemoryUser를 꺼내서 내부의 값을 가져오는 방식에서 JwtMemoryUserDetails에서 값을 제공 받는 방식으로 변경하였습니다.

related to #12

CokeLee777 committed:
refactor: refactoring token domain creater, service, provider

토큰 관련 로직 구현

related to #12

JSY8869 committed:
refactor: Refactoring variable name and annotation

1. 변수 명을 통일 시키기 위해 일관성 있게 수정
RefreshToken -> R
AccessToken -> A

2. 불필요한 주석 제거
문제가 해결되었지만 한 곳에만 불필요하게 남아있던 주석을 제거하였습니다.

related to #12

CokeLee777 reviewed (commented):
제네릭에 대한 네이밍 컨벤션이 보통 저런식으로 하나요?

JSY8869 commented:
E - Element (used extensively by the Java Collections Framework)
K - Key
N - Number
T - Type
V - Value
S,U,V etc. - 2nd, 3rd, 4th types

찾아보니 보통 이렇게 한다고 하네요
수정 하고 커밋 하겠습니다!

JSY8869 committed:
Revert "refactor: Refactoring variable name and annotation"

This reverts commit a32e97882432ee02305444b503a2d603b209bbd1.

JSY8869 committed:
refactor: Refactoring variable name and annotation

1. 일반적인 제네릭 네이밍 컨벤션에 맞춰 수정
제네릭의 네이밍을 통일시키기 위해 일반적인 컨벤션을 기준으로 수정하였습니다.

2. 불필요한 주석 제거
문제가 해결되었지만 한 곳에만 불필요하게 남아있던 주석을 제거하였습니다.

related to #12

CokeLee777 reviewed (approved):
확인했습니다~ 정리해서 Merge 해주세요
